libpam/pam_modutil_getlogin.c

```diff
@@ -19,62 +18,33 @@ pam_modutil_getlogin(pam_handle_t *pamh)
 {
     int status;
     const void *logname;
-    const void *void_curr_tty;
-    const char *curr_tty;
     char *curr_user;
-    struct utmp *ut, line;
+    size_t curr_user_len;
 
     status = pam_get_data(pamh, _PAMMODUTIL_GETLOGIN, &logname);
     if (status == PAM_SUCCESS) {
 	return logname;
     }
 
-    status = pam_get_item(pamh, PAM_TTY, &void_curr_tty);
-    if ((status != PAM_SUCCESS) || (void_curr_tty == NULL))
-      curr_tty = ttyname(0);
-    else
-      curr_tty = (const char*)void_curr_tty;
-
-    if (curr_tty == NULL) {
-	return NULL;
-    }
-
-    if (curr_tty[0] == '/') {   /* full path */
-        const char *t;
-	curr_tty++;
-	if ((t = strchr(curr_tty, '/')) != NULL) {
-	  curr_tty = t + 1;
-	}
+    logname = getlogin();
+    if (logname == NULL) {
+      return NULL;
     }
-    logname = NULL;
 
-    setutent();
-    strncpy(line.ut_line, curr_tty, sizeof(line.ut_line));
-
-    if ((ut = getutline(&line)) == NULL) {
-	goto clean_up_and_go_home;
-    }
-
-    curr_user = calloc(sizeof(line.ut_user)+1, 1);
+    curr_user_len = strlen(logname)+1;
+    curr_user = calloc(curr_user_len, 1);
     if (curr_user == NULL) {
-	goto clean_up_and_go_home;
+      return NULL;
     }
 
-    strncpy(curr_user, ut->ut_user, sizeof(ut->ut_user));
-    /* calloc already zeroed the memory */
+    strncpy(curr_user, logname, curr_user_len);
```

Comment:
The compiler seems to be too picky about this, let's use memcpy instead.

Reply:
Not only too picky, but also wrong...